# Peer-graded Assignment: Build a Regression Model in Keras

## Concrete Strength Prediction - Part D

### Submitted by: Muhammad Jawad Bashir

**1. Importing Libararies & Dataset**

In [1]:
import pandas as pd
import numpy as np

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

**2. Separate into features & labels**

In [4]:
#Let's split the data into predictors and target
cd_columns = concrete_data.columns
predictors = concrete_data[cd_columns[cd_columns != 'Strength']]
target = concrete_data['Strength']
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

**2a: Normalizing the Dataset***

In [5]:
pred_norm = (predictors - predictors.mean()) / predictors.std()
pred_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [6]:
inp_cols = pred_norm.shape[1]
inp_cols

8

**3. Defining Keras Regression Model (Part D: 3 Hidden Layers)**

In [7]:
import keras
from keras.models import Sequential
from keras.layers import Dense

In [9]:
def model_reg():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(inp_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    model.compile(optimizer = 'adam', loss='mean_squared_error')
    return model

**4. Running the model for 50 times with given requirements in Assignment Part-D (3 Hidden Layers with Part-B requirements)**

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [11]:
mse_list = []
model = model_reg()
for i in range(0,50):
    X_train, X_test, y_train, y_test = train_test_split(pred_norm, target, test_size = 0.30, random_state=42)
    model.fit(X_train, y_train, epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test,y_pred)
    mse_list.append(mse)

In [12]:
mse_list

[124.01966802317926,
 101.94877761636349,
 83.80959560433553,
 68.35547834750687,
 57.40544012087256,
 49.32549852674571,
 44.07071685577032,
 45.072958172981934,
 43.167670003351425,
 43.27003191425577,
 42.60245601055456,
 42.1761023690628,
 41.721333638166314,
 43.61200663068037,
 41.793154006103215,
 42.21696901843365,
 42.12170773874763,
 39.731983835735704,
 39.002490342966446,
 39.440671258070154,
 39.96619855473058,
 39.50211562521592,
 38.596009365461455,
 41.231970242754706,
 38.219282948634074,
 37.91658995858758,
 39.41493366770894,
 37.116831932730165,
 37.260355257038306,
 36.37831180598228,
 36.13374695855022,
 35.88613499605027,
 36.05944648917973,
 36.70816422883643,
 35.723233185067244,
 35.45942906558889,
 36.698142078076536,
 37.33712230331601,
 36.433393060312824,
 37.0912450112373,
 36.54722560093452,
 36.43050804808438,
 35.70398367620784,
 37.03764603861482,
 40.76610635196908,
 38.40309525185604,
 36.915910238779524,
 36.30645158463046,
 36.53737414888834,
 36.

**5. MSE mean & Standard Deviation for Task-D**

In [13]:
print('Mean for Task-D: {}'.format(np.mean(mse_list)))
print('Standard Deviation for Task-D: {}'.format(np.std(mse_list)))

Mean for Task-D: 43.89644996252377
Standard Deviation for Task-D: 16.51500318270242


### Difference between MSEs of Part-B (1 Hidden Layer) and Part-D(3 Hidden Layers)

MSE in Part B = 50.019

MSE in Part C = 43.896

MSE decreased by 6.123 after increasing the number of hidden layers to 3 in Part-D.
Hence, it shows that adding more layers decreased the MSE more than by just increasing the number of epochs as we did in
part-C.